File: ports/node/bastionkit.test.mjs
// Smoke test for the bastionkit Node port. Standard library only (node:test,
// node:assert). Offline; no network. Run: node --test ports/node/
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  CONTROLS, TOOL_NAME, baselineDocuments, generateBundle, q, main,
} from "./bastionkit.mjs";

test("tool name", () => {
  assert.equal(TOOL_NAME, "bastionkit");
});

test("control catalogue has all 16 controls", () => {
  assert.equal(CONTROLS.length, 16);
});

test("every control has 4 fields and a known severity", () => {
  const sevs = new Set(["critical", "high", "medium", "low", "info"]);
  for (const c of CONTROLS) {
    assert.equal(c.length, 4);
    assert.ok(sevs.has(c[1]), c[0]);
  }
});

test("scalar quoting", () => {
  assert.equal(q(true), "true");
  assert.equal(q(false), "false");
  assert.equal(q(50), "50");
  assert.equal(q("restricted"), "restricted");
  assert.equal(q("true"), '"true"');
});

test("baseline documents include the required kinds", () => {
  const kinds = new Set(baselineDocuments("prod").map((d) => d.kind));
  for (const k of ["Namespace", "NetworkPolicy", "ResourceQuota",
                   "LimitRange", "Role", "RoleBinding", "ServiceAccount"]) {
    assert.ok(kinds.has(k), `missing ${k}`);
  }
});

test("default-deny netpol selects all pods", () => {
  const dd = baselineDocuments("prod").find(
    (d) => d.kind === "NetworkPolicy" && d.metadata.name === "default-deny-all");
  assert.deepEqual(dd.spec.podSelector, {});
  assert.deepEqual(dd.spec.policyTypes, ["Ingress", "Egress"]);
});

test("service account disables token automount", () => {
  const sa = baselineDocuments("prod").find((d) => d.kind === "ServiceAccount");
  assert.equal(sa.automountServiceAccountToken, false);
});

test("namespace labelled restricted", () => {
  const ns = baselineDocuments("prod").find((d) => d.kind === "Namespace");
  assert.equal(ns.metadata.labels["pod-security.kubernetes.io/enforce"], "restricted");
});

test("custom pod quota propagates", () => {
  const rq = baselineDocuments("p", { podQuota: 7 }).find(
    (d) => d.kind === "ResourceQuota");
  assert.equal(rq.spec.hard.pods, 7);
});

test("generated bundle is multi-document yaml", () => {
  const text = generateBundle("web");
  assert.ok(text.startsWith("---\n"));
  assert.ok(text.includes("kind: NetworkPolicy"));
  assert.ok(text.includes("pod-security.kubernetes.io/enforce: restricted"));
});

test("baseline command exits zero", () => {
  assert.equal(main(["baseline"]), 0);
});

test("generate without namespace exits 2", () => {
  assert.equal(main(["generate"]), 2);
});

File: ports/go/bastionkit_test.go
package main

import (
	"strings"
	"testing"
)

func TestControlCount(t *testing.T) {
	if len(Controls) != 16 {
		t.Fatalf("want 16 controls, got %d", len(Controls))
	}
}

func TestEverySeverityValid(t *testing.T) {
	valid := map[string]bool{"critical": true, "high": true, "medium": true, "low": true, "info": true}
	for _, c := range Controls {
		if !valid[c.Severity] {
			t.Errorf("control %s has invalid severity %q", c.Rule, c.Severity)
		}
		if c.Title == "" || c.Guidance == "" {
			t.Errorf("control %s missing title/guidance", c.Rule)
		}
	}
}

func TestRuleIDsUnique(t *testing.T) {
	seen := map[string]bool{}
	for _, c := range Controls {
		if seen[c.Rule] {
			t.Errorf("duplicate rule id %s", c.Rule)
		}
		seen[c.Rule] = true
	}
}

func TestGenerateContainsBaselineKinds(t *testing.T) {
	out := GenerateBundle("prod", QuotaOpts{})
	for _, k := range []string{"kind: Namespace", "kind: NetworkPolicy",
		"kind: ResourceQuota", "kind: ServiceAccount", "kind: Role"} {
		if !strings.Contains(out, k) {
			t.Errorf("bundle missing %q", k)
		}
	}
}

func TestGenerateRestrictedLabel(t *testing.T) {
	out := GenerateBundle("prod", QuotaOpts{})
	if !strings.Contains(out, "pod-security.kubernetes.io/enforce: restricted") {
		t.Error("bundle does not enforce restricted PodSecurity")
	}
}

func TestGenerateDefaultDeny(t *testing.T) {
	out := GenerateBundle("prod", QuotaOpts{})
	if !strings.Contains(out, "name: default-deny-all") {
		t.Error("bundle missing default-deny NetworkPolicy")
	}
	if !strings.Contains(out, "podSelector: {}") {
		t.Error("default-deny must select all pods")
	}
}

func TestGenerateAutomountDisabled(t *testing.T) {
	out := GenerateBundle("prod", QuotaOpts{})
	if !strings.Contains(out, "automountServiceAccountToken: false") {
		t.Error("service account must disable token automount")
	}
}

func TestGenerateCustomPodQuota(t *testing.T) {
	out := GenerateBundle("prod", QuotaOpts{Pods: 7})
	if !strings.Contains(out, "pods: 7") {
		t.Error("custom pod quota not applied")
	}
}

func TestGenerateNamespacePropagates(t *testing.T) {
	out := GenerateBundle("payments", QuotaOpts{})
	if !strings.Contains(out, "namespace: payments") {
		t.Error("namespace not propagated into bundle")
	}
}

func TestRunBaselineExitsZero(t *testing.T) {
	if rc := run([]string{"baseline"}); rc != 0 {
		t.Errorf("baseline exit = %d, want 0", rc)
	}
}

func TestRunGenerateRequiresNamespace(t *testing.T) {
	if rc := run([]string{"generate"}); rc != 2 {
		t.Errorf("generate without --namespace exit = %d, want 2", rc)
	}
}

func TestRunVersion(t *testing.T) {
	if rc := run([]string{"--version"}); rc != 0 {
		t.Errorf("--version exit = %d, want 0", rc)
	}
}
